### PR TITLE
feat: OIDC now conditionally overwrites the domain for the user

### PR DESCRIPTION
**Pull request by indietyp**
Fixes https://github.com/outline/outline/issues/2542#issuecomment-917324993 and #2678

The problem is that the subdomain of the email is used, which is not always desirable, especially in self-hosted circumstances where the same team might use different email domains (e.g., example.com, extern.example.com, intern.example.com). The current behavior puts everyone into different teams even tho they have the same top-level domain.

The changes here overwrite the domain used, **after** all checks are done on the email domain, ensuring that `allowedDomains` are still checked. This way, domains can still be restricted if needed, but everyone is implicitly put into the same team.

### CODE REVIEW

**Comment by indietyp** [comment]
Hi @tommoor,

I would be willing to implement such behavior, but when setting `ALLOWED_DOMAINS`, which domain should be used as the team name? I used the new OICD to circumvent this problem.

One solution would be to take the first domain from `ALLOWED_DOMAINS` and use that one, but that is very implicit.

Should this functionality be to the `DEPLOYMENT` env so that it won't happen for the cloud installation?

**Comment by tommoor** [comment]
> which domain should be used as the team name?

Which team name are you thinking of? It could probably use the first team that signs up, it's only a default and is all configurable in the in-app settings anyway.

> Should this functionality be to the DEPLOYMENT env so that it won't happen for the cloud installation?

Yes, should be flagged behind `DEPLOYMENT !== "hosted"`

**Comment by indietyp** [comment]
I moved from the authentication to the `accountProvisioner`, is this better?

File: server/commands/accountProvisioner.js [hunk 1]
```diff
@@ -53,13 +53,15 @@ export default async function accountProvisioner({
...
   try {
-    result = await teamCreator({
-      name: teamParams.name,
-      domain: teamParams.domain,
-      subdomain: teamParams.subdomain,
-      avatarUrl: teamParams.avatarUrl,
-      authenticationProvider: authenticationProviderParams,
-    });
+    result =
+      (await findExistingTeam(authenticationProviderParams)) ||
```
Comment (tommoor) [inline server/commands/accountProvisioner.js:57]: Move this to the top of `teamCreator` rather than exporting and using here and I think we're good – are you able to add some basic tests around these expectations?

Comment (indietyp) [reply server/commands/accountProvisioner.js:57]: sure can do!

Comment (kristianvld) [reply server/commands/accountProvisioner.js:57]: @indietyp How is the progress on these changes?

Comment (indietyp) [reply server/commands/accountProvisioner.js:57]: Sorry, I completely forgot to implement these changes. Our company moved on from evaluating Outline so this was subsequently shifted out of my focus and into my freetime. I am hoping to be able to finish the review next Tuesday.

**Comment by almereyda** [comment]
Hi @indietyp !

We have been testing this branch with @tg-x, where we both have a different email domain set in our OIDC IDP, and it did not allow them to log in to an instance configured for "my" email domain, the first one used to initialise the Wiki team.

Is there anything else we can test, before this is refactored as requested above and subsequently merged?

How can we make sure that this feature branch solves #2678?

Thank you for your effort and contribution here.

**Comment by indietyp** [comment]
Hi, I am terribly sorry for the delay. Thank you for looking into it. 

I will try to rewrite and test the pull request this weekend. With university and work, I have been a bit overloaded and didn't really have enough free time to focus on open-source much.

**Comment by almereyda** [comment]
Bilal,

thank you very much for your effort to help us self-hosters with opening
Ouline to a wider user base.
And thanks to Tom and the rest of the team for releasing their core
business asset to the world as open source, which allows us to use and
improve it.

Please also take your time and don't feel pressed by us, especially in this
last, shorter month of the year.

Take it easy,

Jon

**Comment by decentral1se** [comment]
Chiming in to say, very much need this to make outline work for my generic openid connect login approach.

Thanks for this work @indietyp :heart:

File: server/commands/accountProvisioner.test.ts [hunk 2]
```diff
@@ -193,4 +194,69 @@ describe("accountProvisioner", () => {
     const collectionCount = await Collection.count();
     expect(collectionCount).toEqual(1);
   });
+
+  it("should skip findExistingTeam in hosted context", async () => {
+    expect(false).toBeTruthy(); // fail on purpose
```
Comment (tommoor) [inline server/commands/accountProvisioner.test.ts:199]: ?

File: server/commands/accountProvisioner.test.ts [hunk 3]
```diff
@@ -193,4 +194,69 @@ describe("accountProvisioner", () => {
...
+    process.env = env; // reset
+  });
+
+  it("should return existing team, in self-hosted context", async () => {
+    const env = process.env;
+    process.env = { ...env, DEPLOYMENT: "xxx" }; // self to something that isn't hosted
+
+    const team = await buildTeam();
+    const authenticationProvider = (await team.getAuthenticationProviders())[0];
+    const existingTeam = await findExistingTeam({
```
Comment (tommoor) [inline server/commands/accountProvisioner.test.ts:221]: This file should be tests for accountProvisioner, we seem to be testing `findExistingTeam` here instead

File: server/commands/accountProvisioner.test.ts [hunk 4]
```diff
@@ -193,4 +194,69 @@ describe("accountProvisioner", () => {
...
+    process.env = env; // reset
+  });
+
+  it("should create new authentication provider on existing team if in self-hosted context", async () => {
+    const env = process.env;
+    process.env = { ...env, DEPLOYMENT: "xxx" }; // self to something that isn't hosted
+
+    const team = await buildTeam(); // initial team accountProvider is slack
+    const providerId = uuidv4();
+    const existingTeam = await findExistingTeam({ name: "OICD", providerId });
```
Comment (tommoor) [inline server/commands/accountProvisioner.test.ts:241]: Similarly to above comment

File: server/commands/accountProvisioner.ts [hunk 5]
```diff
@@ -45,6 +45,44 @@ export type AccountProvisionerResult = {
   isNewUser: boolean;
 };
 
+export async function findExistingTeam(
+  authenticationProvider: Props["authenticationProvider"]
+): Promise<TeamCreatorResult | null> {
+  // Should outline deployed in a multi-tenant environment, skip searching
+  // for an existing team.
+  console.log(process.env.DEVELOPMENT);
```
Comment (tommoor) [inline server/commands/accountProvisioner.ts:53]: Logging

File: server/commands/accountProvisioner.ts [hunk 6]
```diff
@@ -45,6 +45,44 @@ export type AccountProvisionerResult = {
...
+  if (process.env.DEPLOYMENT === "hosted") return null;
+
+  // get the first team that exists, ordered by createdAt
+  const team = await Team.findOne({ limit: 1, order: ["createdAt"] });
+  if (team === null) {
+    return null;
+  }
+
+  // query if a corresponding authenticationProvider already exists
+  const authenticationProviders = await team.getAuthenticationProviders({
```
Comment (tommoor) [inline server/commands/accountProvisioner.ts:63]: This logic doesn't seem to be restricted to OIDC in any way, which means if someone tries to signin with another Google account then it would just remap them to an existing one and allow signin?

File: server/commands/accountProvisioner.ts [hunk 7]
```diff
@@ -55,13 +93,15 @@ export default async function accountProvisioner({
...
   try {
-    result = await teamCreator({
-      name: teamParams.name,
-      domain: teamParams.domain,
-      subdomain: teamParams.subdomain,
-      avatarUrl: teamParams.avatarUrl,
-      authenticationProvider: authenticationProviderParams,
-    });
+    result =
+      (await findExistingTeam(authenticationProviderParams)) ||
```
Comment (tommoor) [inline server/commands/accountProvisioner.ts:97]: This should be part of teamCreator instead of it's own thing

File: package.json [hunk 8]
```diff
@@ -238,7 +238,7 @@
     "@types/tmp": "^0.2.2",
     "@types/turndown": "^5.0.1",
     "@types/utf8": "^3.0.0",
-    "@typescript-eslint/eslint-plugin": "^5.3.1",
+    "@typescript-eslint/eslint-plugin": "^5.6.0",
```
Comment (tommoor) [inline package.json:241]: Unrelated package changes seem to have introduced a lot of yarn.lock changes.

**Review by tommoor — CHANGES_REQUESTED** [review]
I appreciate the work here however as it stands the PR is still far off from what would be landable, it also seems to introduce a large security hole in the form of re-mapping auth providers if an existing auth provider of the same name exists.

It's a shame that the domain was mapped from the email in the original OIDC implementation – we should probably have just had it as a static string, but changing that now would break existing installations.

I'm going to try and tackle this as part of  https://github.com/outline/outline/pull/2986
